FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1457 @@
+
+    for( i = 0; i < SUBSCRIPTIONS_MAX_COUNT; i++ )
+    {
+        if( pxSubscriptions[ i ].topicFilterLength == topicFilterLength )
+        {
+            if( ( strncmp( pxSubscriptions[ i ].pTopicFilter, pTopicFilter, topicFilterLength ) == 0 ) && true )
+            {
+                pxSubscriptions[ i ].topicFilterLength = 0;
+                pxSubscriptions[ i ].pResponseQueue = NULL;
+                memset( ( void * ) pxSubscriptions[ i ].pTopicFilter, 0x00, sizeof( pxSubscriptions[ i ].pTopicFilter ) );
+                break;
+            }
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvCreateCommand( CommandType_t xCommandType,
+                              CommandContext_t * pxContext,
+                              CommandCallback_t xCallback,
+                              Command_t * pxCommand )
+{
+    bool xIsValid = true;
+
+    memset( ( void * ) pxCommand, 0x00, sizeof( Command_t ) );
+    pxCommand->commandType = xCommandType;
+    pxCommand->pContext = pxContext;
+    pxCommand->callback = xCallback;
+
+    /* Determine if required parameters are present in context. */
+    switch( xCommandType )
+    {
+        case PUBLISH:
+            xIsValid = ( pxContext != NULL ) ? ( pxContext->pPublishInfo != NULL ) : false;
+            break;
+
+        case SUBSCRIBE:
+        case UNSUBSCRIBE:
+            xIsValid = ( pxContext != NULL ) ? ( pxContext->pSubscribeInfo != NULL ) : false;
+            break;
+
+        default:
+            /* Other operations don't need a context. */
+            break;
+    }
+
+    return xIsValid;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvAddCommandToQueue( Command_t * pxCommand )
+{
+    xQueueSend( xCommandQueue, pxCommand, DEMO_TICKS_TO_WAIT );
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvCopyPublishToQueue( MQTTPublishInfo_t * pPublishInfo,
+                                   QueueHandle_t pResponseQueue )
+{
+    PublishElement_t xCopiedPublish;
+    MQTTPublishInfo_t * pxCopiedPublishInfo = NULL;
+
+    memset( ( void * ) &xCopiedPublish, 0x00, sizeof( xCopiedPublish ) );
+    pxCopiedPublishInfo = &( xCopiedPublish.publishInfo );
+    memcpy( &( xCopiedPublish.publishInfo ), pPublishInfo, sizeof( MQTTPublishInfo_t ) );
+
+    /* Since adding an MQTTPublishInfo_t to a queue will not copy its string buffers,
+     * we need to add buffers to a struct and copy the entire structure. */
+    memcpy( xCopiedPublish.pTopicName, pPublishInfo->pTopicName, pPublishInfo->topicNameLength );
+    memcpy( xCopiedPublish.pPayload, pPublishInfo->pPayload, pPublishInfo->payloadLength );
+    pxCopiedPublishInfo->pTopicName = ( const char * ) xCopiedPublish.pTopicName;
+    pxCopiedPublishInfo->pPayload = xCopiedPublish.pPayload;
+
+    /* Add to response queue. */
+    xQueueSendToBack( pResponseQueue, pxCopiedPublishInfo, DEMO_TICKS_TO_WAIT );
```

Comment:
You are sending the MQTTPublishInfo_t to the queue and it will copy over only pointers to pTopicName and pPayload. These pointers are allocated in stack. Is that what you want to do? Do you want to send xCopiedPublish instead?

Reply:
Good catch, yes that's exactly what I wanted to do. I'll change it. I expect this happened to work as originally intended due to the fact that the publish info is the first member of the PublishElement_t struct, so the pointer to it should be the same as the pointer to the struct.